### DISCUSSION TITLE
TypeError when passing loop as first argument into `React\Socket\Server`

### DISCUSSION THREAD

**Discussion by eric-lukyamuzi** [reactphp/reactphp#523] [Q&A] [answered]
Hi everyone,

thankyou for the great work done on this library.

its my first time using it, but am having this error below please help me.

```
PHP Fatal error: Uncaught TypeError: strpos(): Argument #1 ($haystack) must be of type string, React\EventLoop\StreamSelectLoop given in /Applications/XAMPP/xamppfiles/htdocs/socket-react/vendor/react/socket/src/Server.php:69
Stack trace:
#0 /Applications/XAMPP/xamppfiles/htdocs/socket-react/vendor/react/socket/src/Server.php(69): strpos(Object(React\EventLoop\StreamSelectLoop), '://')
#1 /Applications/XAMPP/xamppfiles/htdocs/socket-react/listen.socket.php(11): React\Socket\Server->__construct(Object(React\EventLoop\StreamSelectLoop))
#2 {main}
thrown in /Applications/XAMPP/xamppfiles/htdocs/socket-react/vendor/react/socket/src/Server.php on line 69
```

*(Updated by @clue to add appropriate code block, please take the time to properly format your comments.)*

**Comment by SimonFrings** [comment, marked as answer]
Hey @eric-lancelot, I can't see the exact code, but it seems like you're trying to pass the event loop as a first argument into `React\Socket\Server` which results in a `TypeError` because the first argument expects a URI of the type string. You can take a look at the `React\Socket\Server` implementation with it's constructor parameters here: https://github.com/reactphp/socket/blob/1.x/src/Server.php#L52

The `React\Socket\Server` is also deprecated since the [`v1.9.0` release of reactphp/socket](https://github.com/reactphp/socket/releases/tag/v1.9.0) and got replaced with [`React\Socket\SocketServer`](https://github.com/reactphp/socket#socketserver). We would always recommend to use the latest  version of this package if possible (current version is v1.13.0).

I can only make assumptions here, but maybe you're installing a very old version of either one of the ReactPHP components or an (old) version of a project using ReactPHP, which installs old ReactPHP versions.

Hope this helps :+1:

FYI: I converted this ticket to a discussion so we give it some more visibility for other users with the same question. I also renamed the title to give it a more precise description of what this ticket is about.

**Reply by eric-lukyamuzi** [reply]
Am using the latest version

Here is my code below

```php
<?php

require 'vendor/autoload.php';

use React\EventLoop\Factory;
use React\Socket\Server;
use React\Socket\ConnectionInterface;

$loop = Factory::create();
$socket = new Server($loop);

$socket->on('connection', function (ConnectionInterface $connection) {
    $connection->on('data', function ($data) use ($connection) {
        // Handle received data from client
        // Example: Echo the received data back to the client
        $connection->write('Received: ' . $data);
    });

    $connection->on('close', function () use ($connection) {
        // Handle client connection close
        // Example: Clean up resources associated with the client
        echo "Connection closed\n";
    });
});

$socket->listen(80, '192.168.0.123'); // Listening on all interfaces, port 2126

echo "Server running...\n";

$loop->run();

?>
```

*(Updated by @clue to add appropriate code block, please take the time to properly format your comments.)*

**Reply by eric-lukyamuzi** [reply]
Hi @clue what am trying to archieve here is to be able to recieve stream data from a GPS and write back data to the device the devices are sending data on port 2122 on my server, 

please help me where am going wrong that code block above.


thankyou in advance

**Reply by SimonFrings** [reply]
The `React\Socket\Server` has the following constructor arguments:
```php
public function __construct($uri, LoopInterface $loop = null, array $context = array())
{
…
```

In your code base you're doing `$socket = new Server($loop);` which is incorrect as the `React\Socket\Server` accepts a `$uri` of type string as its first parameter and an optional `$loop` as its second parameter. So what you have to do here is:

```php
$uri = '127.0.0.1:8080'; // Just an example uri, adapt value to your needs
$loop = Factory::create();
$socket = new Server($uri, $loop);
```

**Reply by eric-lukyamuzi** [reply]
it has worked thankyou.

**Reply by eric-lukyamuzi** [reply]
```
$connection->on('data', function ($data) use ($connection) {
       
        $connection->write('Received: ' . $data);
    });
```

**the variable $data is returning me nothing**

Full code base below :

```
<?php


require 'vendor/autoload.php';

use React\EventLoop\Factory;
use React\Socket\Server;
use React\Socket\ConnectionInterface;

$uri = 'myip:2126';
$loop = Factory::create();
$socket = new Server($uri, $loop);

$socket->on('connection', function (ConnectionInterface $connection) {
    $connection->on('data', function ($data) use ($connection) {
       
        $connection->write('Received: ' . $data);
    });

    $connection->on('close', function () use ($connection) {
    
        echo "Connection closed\n";
    });
});

//$socket->listen(80, '192.168.0.123');

echo "Server running...\n";

$loop->run();



?>
```

**Reply by eric-lukyamuzi** [reply]
i can see that the GPS devices are trying to send data but the variable $data does not print it( see attached screenshot )
<img width="212" alt="Screenshot 2023-07-10 at 3 38 52 PM" src="https://github.com/reactphp/reactphp/assets/58414055/ff308637-7750-4bf7-8e5f-bf1ff353df02">

**Reply by SimonFrings** [reply]
> i can see that the GPS devices are trying to send data but the variable $data does not print it( see attached screenshot )

@eric-lancelot  In your screenshot you're looking at the messages on the server side, for example, every time a connection closes you do a `echo "Connection closed\n";` and can see the output in your console. Once you receive data from one of your clients, you use the `write()` method to send a message back to your client, but you never echo `$data` on the server side. To print the received data, you can do something like this:
```php
…

$socket->on('connection', function (ConnectionInterface $connection) {
    $connection->on('data', function ($data) use ($connection) {
       
        echo "Data: " . $data . "\n";
        $connection->write('Received: ' . $data);
    });

    $connection->on('close', function () use ($connection) {
    
        echo "Connection closed\n";
    });
});

…
```
Since you mentioned that you're working with GPS devices, it's unclear whether the content of your data is in JSON format or perhaps something binary. Regardless, I'm confident that you'll find the appropriate parsing methods to suit your needs ;)

Just one more note here: You should use the new [`React\Socket\SocketServer`](https://github.com/reactphp/socket#socketserver) instead of the deprecated `React\Socket\Server`. We're working towards a new major version of ReactPHP which will get rid of old & deprecated APIs, so `React\Socket\Server` will be removed at some point in the future.

We also introduced the ReactPHP default loop with [reactphp/event-loop v1.2.0](https://github.com/reactphp/event-loop/releases/tag/v1.2.0), so you don't have to pass a `$loop` variable into SocketServer instance as it will use the same loop by default anyway. Here's what your code should look like with our latest APIs:

```php
<?php


require 'vendor/autoload.php';

use React\Socket\SocketServer;
use React\Socket\ConnectionInterface;

$uri = 'myip:2126';
$socket = new SocketServer($uri);

$socket->on('connection', function (ConnectionInterface $connection) {
    $connection->on('data', function ($data) use ($connection) {
       
        $connection->write('Received: ' . $data);
    });

    $connection->on('close', function () use ($connection) {
    
        echo "Connection closed\n";
    });
});

echo "Server running...\n";
```
